Server/Components/Devices/DevicesFrame.razor.cs

```diff
@@ -50,24 +54,33 @@ public partial class DevicesFrame : AuthComponentBase, IDisposable
     private IDataService DataService { get; init; } = null!;
 
     [Inject]
-    private IJsInterop JsInterop { get; init; } = null!;
+    private IToastService ToastService { get; init; } = null!;
 
     [Inject]
-    private IToastService ToastService { get; init; } = null!;
+    private IMessenger Messenger { get; init; } = null!;
 
     private int TotalPages => (int)Math.Max(1, Math.Ceiling((decimal)_filteredDevices.Count / _devicesPerPage));
 
     public void Dispose()
     {
-        CircuitConnection.MessageReceived -= CircuitConnection_MessageReceived;
-        AppState.PropertyChanged -= AppState_PropertyChanged;
+        Messenger.Unregister<DisplayNotificationMessage, string>(this, CircuitConnection.ConnectionId);
+        Messenger.Unregister<ScriptResultMessage, string>(this, CircuitConnection.ConnectionId);
+        Messenger.Unregister<DeviceStateChangedMessage, string>(this, CircuitConnection.ConnectionId);
+
         GC.SuppressFinalize(this);
     }
 
-    public void Refresh()
+    private async Task HandleDisplayNotificationMessage(DisplayNotificationMessage message)
+    {
+        TerminalStore.AddTerminalLine(message.ConsoleText);
+        ToastService.ShowToast(message.ToastText, classString: message.ClassName);
+        await InvokeAsync(StateHasChanged);
+    }
+
+    public async Task Refresh()
```

Comment:
Just curious, how would cancellation work here if you wanted to pass a token?

Reply:
We'd probably want to grab it from the overridden `CircuitHandler.OnCircuitOpenedAsync`.  I don't know if this token would be cancelled when the app shuts down.  If not, we'd probably want to combine it with `IHostApplication.ApplicationStopping`.

![image](https://github.com/immense/Remotely/assets/20995508/f7f4e2b3-c99d-4a96-bf9a-3af8ebf4e7df)